File: libs/ui_pages_mapedit/src/map_topology_issue_panel.cpp
#include "pwb/ui_pages_mapedit/map_topology_issue_panel.hpp"

#include "pwb/ui_widgets/object_table.hpp"
#include "pwb/ui_widgets/states.hpp"

#include <QLabel>
#include <QResizeEvent>
#include <QVBoxLayout>

namespace pwb::ui_pages_mapedit {

// --- TopologyCellProxy --------------------------------------------------------

QString TopologyCellProxy::text() const {
    const QVariant value =
        model_->data(model_->index(row_, column_));
    return value.isNull() ? QString() : value.toString();
}

// --- TopologyTableView --------------------------------------------------------

TopologyTableView::TopologyTableView(QWidget* parent) : QTableView(parent) {}

int TopologyTableView::rowCount() const {
    QAbstractItemModel* m = model();
    return m == nullptr ? 0 : m->rowCount();
}

TopologyCellProxy* TopologyTableView::item(int row, int column) {
    QAbstractItemModel* m = model();
    if (m == nullptr || row < 0 || column < 0 || row >= m->rowCount() ||
        column >= m->columnCount()) {
        return nullptr;
    }
    last_item_.emplace(m, row, column);
    return &*last_item_;
}

void TopologyTableView::setCurrentCell(int row, int column) {
    QAbstractItemModel* m = model();
    if (m == nullptr) {
        return;
    }
    setCurrentIndex(m->index(row, column));
}

void TopologyTableView::emit_item_double_clicked(const QModelIndex& index) {
    emit itemDoubleClicked(item(index.row(), index.column()));
}

// --- MapTopologyIssuePanel -----------------------------------------------------

MapTopologyIssuePanel::MapTopologyIssuePanel(QWidget* parent)
    : QWidget(parent) {
    auto* layout = new QVBoxLayout(this);
    summary = new QLabel(QStringLiteral("当前没有拓扑问题"), this);
    // Rows are packed as (key, issue-map) pairs — Python ``pair[1].get()``.
    const auto issue_of = [](const QVariant& row) {
        const auto pair = row.toList();
        return pair.size() >= 2 ? pair[1].toMap() : QVariantMap();
    };
    // Python ``str(issue.get(key, "") or "")``: falsy scalars (None, 0,
    // False, "") all collapse to "".
    const auto truthy_cell = [](const QVariant& v) -> QString {
        if (v.isNull()) {
            return {};
        }
        switch (v.typeId()) {
            case QMetaType::Bool:
                return v.toBool() ? QStringLiteral("True") : QString();
            case QMetaType::Int:
            case QMetaType::LongLong:
            case QMetaType::UInt:
            case QMetaType::ULongLong:
            case QMetaType::Double:
                return v.toDouble() == 0.0 ? QString() : v.toString();
            default:
                return v.toString();
        }
    };
    model_ = new ui_widgets::ObjectTableModel(
        {ui_widgets::ColumnSpec{
             QStringLiteral("feature_id"), QStringLiteral("要素"),
             [issue_of, truthy_cell](const QVariant& row) {
                 return truthy_cell(
                     issue_of(row).value(QStringLiteral("feature_id")));
             }},
         ui_widgets::ColumnSpec{
             QStringLiteral("message"), QStringLiteral("问题"),
             [issue_of, truthy_cell](const QVariant& row) {
                 return truthy_cell(
                     issue_of(row).value(QStringLiteral("message")));
             }},
         ui_widgets::ColumnSpec{
             QStringLiteral("severity"), QStringLiteral("级别"),
             [issue_of, truthy_cell](const QVariant& row) {
                 return truthy_cell(
                     issue_of(row).value(QStringLiteral("severity")));
             }}},
        [](const QVariant& row) {
            const auto pair = row.toList();
            return pair.isEmpty() ? QString() : pair.first().toString();
        },
        this);
    table = new TopologyTableView(this);
    table->setModel(model_);
    ui_widgets::bind_table_defaults(table);
    layout->addWidget(summary);
    layout->addWidget(table);

    empty_state_ = new ui_widgets::PwbEmptyState(
        QStringLiteral("未发现拓扑问题"),
        QStringLiteral("图幅拓扑检查通过。"), QStringLiteral("inbox.svg"),
        nullptr, this);
    empty_state_->setAttribute(Qt::WA_TransparentForMouseEvents);
    empty_state_->setParent(table);
    empty_state_->hide();
    connect(model_, &QAbstractItemModel::modelReset, this,
            [this]() { update_empty_state(); });
    connect(model_, &QAbstractItemModel::rowsInserted, this,
            [this](const QModelIndex&, int, int) { update_empty_state(); });
    connect(model_, &QAbstractItemModel::rowsRemoved, this,
            [this](const QModelIndex&, int, int) { update_empty_state(); });

    connect(table, &QTableView::activated, this,
            [this](const QModelIndex& index) { on_row_activated(index); });
    connect(table, &QTableView::doubleClicked, table,
            [this](const QModelIndex& index) {
                table->emit_item_double_clicked(index);
            });
    connect(table, &TopologyTableView::itemDoubleClicked, this,
            [this](TopologyCellProxy* item) {
                on_item_double_clicked(item);
            });
    update_empty_state();
}

void MapTopologyIssuePanel::on_item_double_clicked(
    TopologyCellProxy* item) {
    if (item != nullptr) {
        on_row_activated(model_->index(item->row(), item->column()));
    }
}

void MapTopologyIssuePanel::set_issues(
    const std::vector<QVariantMap>& issues) {
    issues_ = issues;
    summary->setText(QStringLiteral("拓扑问题：%1").arg(issues.size()));
    // 行对象包一层 (稳定键, issue)：issue dict 无业务 id，键由确定性
    // 序号+要素 id 构成，重复 set_issues 走 set_rows 差分路径。
    std::vector<QVariant> rows;
    rows.reserve(issues_.size());
    int i = 0;
    for (const auto& issue : issues_) {
        const QString key =
            QStringLiteral("%1:%2")
                .arg(i++)
                .arg(issue.value(QStringLiteral("feature_id")).toString());
        rows.push_back(QVariantList{key, issue});
    }
    model_->set_rows(rows);
}

void MapTopologyIssuePanel::on_row_activated(const QModelIndex& index) {
    // Double-click / Enter on a row asks to locate its feature. The row
    // OBJECT is read through the model — a sorted view reorders display
    // rows, so issues_[row] would locate the wrong feature.
    const QVariant row = model_->row_at(index.row());
    const auto pair = row.toList();
    if (pair.size() < 2) {
        return;
    }
    const QString feature_id =
        pair[1].toMap().value(QStringLiteral("feature_id")).toString();
    if (!feature_id.isEmpty()) {
        emit locate_requested(feature_id);
    }
}

void MapTopologyIssuePanel::resizeEvent(QResizeEvent* event) {
    QWidget::resizeEvent(event);
    if (empty_state_->isVisible()) {
        empty_state_->setGeometry(table->viewport()->rect());
    }
}

void MapTopologyIssuePanel::update_empty_state() {
    if (model_->rowCount() == 0) {
        empty_state_->setGeometry(table->viewport()->rect());
        empty_state_->show();
        empty_state_->raise();
    } else {
        empty_state_->hide();
    }
}

}  // namespace pwb::ui_pages_mapedit

File: legacy/python_reference/product/paleo_workbench/ui/pages/map_topology_issue_panel.py
"""编图工作台底部「拓扑问题」面板。

V11 Model/View 迁移（goal §8 / 01-ui-audit D2）：问题表改
``QTableView + ObjectTableModel``——拓扑检查可产生无上限条目，此前
setItemCount+setItem 逐格重建；行为对象为 issue dict，列取值即
``issue.get(key)``。无问题时表上覆盖统一空态（词汇与任务中心一致）。
"""

from __future__ import annotations

from PySide6.QtCore import QModelIndex, Qt, Signal
from PySide6.QtWidgets import QLabel, QTableView, QVBoxLayout, QWidget

from paleo_workbench.ui.components.states import PwbEmptyState
from paleo_workbench.ui.modelview import (
    ColumnSpec,
    ObjectTableModel,
    bind_table_defaults,
)


class _TopologyCellProxy:
    """Test-facing cell handle（row/column/text），无 QTableWidgetItem。"""

    def __init__(self, model, row: int, column: int) -> None:
        self._model = model
        self._row = row
        self._column = column

    def row(self) -> int:
        return self._row

    def column(self) -> int:
        return self._column

    def text(self) -> str:
        value = self._model.data(self._model.index(self._row, self._column))
        return "" if value is None else str(value)


class _TopologyTableView(QTableView):
    """QTableView + QTableWidget 兼容访问器（双击信号/单元格定位）。"""

    # QTableWidget 兼容信号：测试与调用方仍按 (row, column)/(item) 语义驱动。
    itemDoubleClicked = Signal(object)

    def rowCount(self) -> int:  # noqa: N802
        model = self.model()
        return 0 if model is None else model.rowCount()

    def item(self, row: int, column: int):
        model = self.model()
        if model is None:
            return None
        if row < 0 or column < 0 or row >= model.rowCount() or column >= model.columnCount():
            return None
        return _TopologyCellProxy(model, row, column)

    def setCurrentCell(self, row: int, column: int) -> None:  # noqa: N802
        model = self.model()
        if model is None:
            return
        self.setCurrentIndex(model.index(row, column))

    def _emit_item_double_clicked(self, index: QModelIndex) -> None:
        self.itemDoubleClicked.emit(self.item(index.row(), index.column()))


class MapTopologyIssuePanel(QWidget):
    locate_requested = Signal(str)

    def __init__(self, parent=None):
        super().__init__(parent)
        layout = QVBoxLayout(self)
        self.summary = QLabel("当前没有拓扑问题")
        self._issues: list[dict] = []
        self._model = ObjectTableModel(
            columns=[
                ColumnSpec("feature_id", "要素", lambda pair: str(pair[1].get("feature_id", "") or "")),
                ColumnSpec("message", "问题", lambda pair: str(pair[1].get("message", "") or "")),
                ColumnSpec("severity", "级别", lambda pair: str(pair[1].get("severity", "") or "")),
            ],
            key_of=lambda pair: pair[0],
            parent=self,
        )
        self.table = _TopologyTableView()
        self.table.setModel(self._model)
        bind_table_defaults(self.table)
        layout.addWidget(self.summary)
        layout.addWidget(self.table)

        self._empty_state = PwbEmptyState("未发现拓扑问题", "图幅拓扑检查通过。", parent=self)
        self._empty_state.setAttribute(Qt.WidgetAttribute.WA_TransparentForMouseEvents)
        self._empty_state.setParent(self.table)
        self._empty_state.hide()
        self._model.modelReset.connect(self._update_empty_state)
        self._model.rowsInserted.connect(self._update_empty_state)
        self._model.rowsRemoved.connect(self._update_empty_state)

        self.table.activated.connect(self._on_row_activated)
        self.table.doubleClicked.connect(self.table._emit_item_double_clicked)
        self.table.itemDoubleClicked.connect(self._on_item_double_clicked)
        self._update_empty_state()

    def _on_item_double_clicked(self, item) -> None:
        if item is not None:
            index = self._model.index(item.row(), item.column())
            self._on_row_activated(index)

    def set_issues(self, issues: list[dict]) -> None:
        self._issues = list(issues)
        self.summary.setText(f"拓扑问题：{len(issues)}")
        # 行对象包一层 (稳定键, issue)：issue dict 无业务 id，键由确定性
        # 序号+要素 id 构成，重复 set_issues 走 set_rows 差分路径。
        self._model.set_rows(
            [
                (f"{i}:{issue.get('feature_id', '')}", issue)
                for i, issue in enumerate(self._issues)
            ]
        )

    def _on_row_activated(self, index: QModelIndex) -> None:
        """Double-click / Enter on a row asks to locate its feature."""
        row = index.row()
        if not 0 <= row < len(self._issues):
            return
        feature_id = str(self._issues[row].get("feature_id", "") or "")
        if feature_id:
            self.locate_requested.emit(feature_id)

    def resizeEvent(self, event) -> None:  # noqa: N802
        super().resizeEvent(event)
        if self._empty_state.isVisible():
            self._empty_state.setGeometry(self.table.viewport().rect())

    def _update_empty_state(self, *_args) -> None:
        if self._model.rowCount() == 0:
            self._empty_state.setGeometry(self.table.viewport().rect())
            self._empty_state.show()
            self._empty_state.raise_()
        else:
            self._empty_state.hide()
